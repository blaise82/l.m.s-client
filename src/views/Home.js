import React, { Component } from 'react';
import classes from '../styles/Home.module.css';
import { BrowserRouter, Link } from 'react-router-dom';
import axios from 'axios';

export default class Home extends Component {
	constructor(props) {
		super(props);
		this.state = {
			email: '',
			emailError: '',
			password: '',
			passwordError: ''
		};
	}
	validate = () => {
		let isError = false;
		const errors = {
			emailError: '',
			passwordError: ''
		};

		if (this.state.email.indexOf('@') === -1) {
			isError = true;
			errors.emailError = 'Requires valid email';
		}
		if (this.state.password.length < 3) {
			isError = true;
			errors.passwordError = 'Please provide a password';
		}
		this.setState({
			...this.state,
			...errors
		});

		return isError;
	};

	onSubmit = async (e) => {
		e.preventDefault();
		const error = await this.validate();
		if (!error) {
			const send = {
				email: this.state.email,
				password: this.state.password
			};

			await axios
				.post(`https://lms-server-octopus.herokuapp.com/api/v1/auth/signin`, send)
				.then((res) => {
					localStorage.setItem('token', res.data.data.token);
					this.setState({
						email: '',
						password: ''
					});
					// redirect comes in here
				})
				.catch((error) => {
					this.setState({
						passwordError: error.response.data.error,
						emailError: error.response.data.error
					});
				});
		}
	};
	change = (e) => {
		this.setState({
			[e.target.name]: e.target.value
		});
	};
	render() {
		return (
			<div>
				<div className={classes.form}>
					<div className={classes.head}>
						<p className={classes.title}>L.M.S - Sign in</p>
						<p className={classes.sub_title}>Please fill the form</p>
					</div>
					<form onSubmit={(e) => this.onSubmit(e)}>
						<ul>
							<li className={classes.input_wrap}>
								<label htmlFor="email" className={classes.label}>
									Email address
								</label>
								<div className={classes.input_dev}>
									<input
										type="text"
										placeholder="Enter email"
										aria-label="email"
										name="email"
										className={classes.input}
										value={this.state.email}
										onChange={(e) => this.change(e)}
									/>
								</div>
								<p className={classes.message} aria-label="emailError">
									{this.state.emailError}
								</p>
							</li>
							<li className={classes.input_wrap}>
								<label htmlFor="email" className={classes.label}>
									Password
								</label>
								<div className={classes.input_dev}>
									<input
										type="password"
										name="password"
										aria-label="password"
										placeholder="Enter password"
										className={classes.input}
										value={this.state.password}
										onChange={(e) => this.change(e)}
									/>
								</div>
								<p className={classes.message} aria-label="passwordError">
									{this.state.passwordError}
								</p>
							</li>
							<div className={classes.action}>
								<BrowserRouter>
									<Link to="/" className={classes.link}>
										Have no account??
									</Link>
								</BrowserRouter>
								<button className={classes.button} aria-label="submit">
									sign in
								</button>
							</div>
						</ul>
					</form>
				</div>
				<div className={classes.one} />
				<div className={classes.two} />
				<div className={classes.three} />
			</div>
		);
	}
}
