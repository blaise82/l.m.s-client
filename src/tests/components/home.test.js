import React, { Component } from 'react';
import { render, fireEvent, cleanup, wait, waitForElement } from '@testing-library/react';
import Home from '../../views/Home';
import userEvent from '@testing-library/user-event';

describe('', () => {
	afterEach(cleanup);
	it('renders the component', () => {
		const { asFragment } = render(<Home />);
		expect(asFragment(<Home />)).toMatchSnapshot();
	});

	it('should show error on submit without providing email and password', async () => {
		const { getByLabelText, getByText, container, debug } = render(<Home />);
		const email = getByLabelText('email');
		const password = getByLabelText('password');
		const submit = getByLabelText('submit');
		const passwordError = getByLabelText('passwordError');
		const emailError = getByLabelText('emailError');

		expect(email.value).toBe('');
		expect(password.value).toBe('');
		await userEvent.click(submit);
		expect(passwordError.textContent).toBe('Please provide a password');
		expect(emailError.textContent).toBe('Requires valid email');
	});

	it('should show error on submit without providing password', async () => {
		const { getByLabelText, getByText, container, debug } = render(<Home />);
		const email = getByLabelText('email');
		const password = getByLabelText('password');
		const submit = getByLabelText('submit');
		const passwordError = getByLabelText('passwordError');
		const emailError = getByLabelText('emailError');

		userEvent.type(email, 'octopusbn@gmail.com');
		expect(email.value).toBe('octopusbn@gmail.com');
		expect(password.value).toBe('');
		await userEvent.click(submit);
		expect(passwordError.textContent).toBe('Please provide a password');

		// console.log(passwordError.textContent);
		// // expect(onSubmit).toHaveBeenCalled()
	});
	it('should show error on submit without providing email and password which are not correct', async () => {
		const { getByLabelText, getAllByText, container, debug } = render(<Home />);
		const email = getByLabelText('email');
		const password = getByLabelText('password');
		const submit = getByLabelText('submit');
		const passwordError = getByLabelText('passwordError');
		const emailError = getByLabelText('emailError');

		userEvent.type(email, 'blaise@gmail.com');
		userEvent.type(password, 'wrong');

		expect(email.value).toBe('blaise@gmail.com');
		expect(password.value).toBe('wrong');
		await fireEvent.click(submit);

		await waitForElement(() => getAllByText('Incorrect username or password combination'));
		expect(emailError.textContent).toBe('Incorrect username or password combination');
		expect(passwordError.textContent).toBe('Incorrect username or password combination');
	});

	it('should submit without providing error', async () => {
		const { getByLabelText, getAllByText, container, debug } = render(<Home />);
		const email = getByLabelText('email');
		const password = getByLabelText('password');
		const submit = getByLabelText('submit');
		const passwordError = getByLabelText('passwordError');
		const emailError = getByLabelText('emailError');

		userEvent.type(email, 'blaise@gmail.com');
		userEvent.type(password, 'password');

		expect(email.value).toBe('blaise@gmail.com');
		expect(password.value).toBe('password');
		await fireEvent.click(submit);

		await wait(() => expect(email.value).toBe(''));
		await wait(() => expect(password.value).toBe(''));
		expect(emailError.textContent).toBe('');
		expect(passwordError.textContent).toBe('');
	});
});
