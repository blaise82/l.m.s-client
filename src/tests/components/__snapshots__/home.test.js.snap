// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the component 1`] = `
<DocumentFragment>
  <div>
    <div
      class="form"
    >
      <div
        class="head"
      >
        <p
          class="title"
        >
          L.M.S - Sign in
        </p>
        <p
          class="sub_title"
        >
          Please fill the form
        </p>
      </div>
      <form>
        <ul>
          <li
            class="input_wrap"
          >
            <label
              class="label"
              for="email"
            >
              Email address
            </label>
            <div
              class="input_dev"
            >
              <input
                aria-label="email"
                class="input"
                name="email"
                placeholder="Enter email"
                type="text"
                value=""
              />
            </div>
            <p
              aria-label="emailError"
              class="message"
            />
          </li>
          <li
            class="input_wrap"
          >
            <label
              class="label"
              for="email"
            >
              Password
            </label>
            <div
              class="input_dev"
            >
              <input
                aria-label="password"
                class="input"
                name="password"
                placeholder="Enter password"
                type="password"
                value=""
              />
            </div>
            <p
              aria-label="passwordError"
              class="message"
            />
          </li>
          <div
            class="action"
          >
            <a
              class="link"
              href="/"
            >
              Have no account??
            </a>
            <button
              aria-label="submit"
              class="button"
            >
              sign in
            </button>
          </div>
        </ul>
      </form>
    </div>
    <div
      class="one"
    />
    <div
      class="two"
    />
    <div
      class="three"
    />
  </div>
</DocumentFragment>
`;
